### PR TITLE
Add allowed authn providers and force users and courses to belong to institutions

### PR DESCRIPTION
**Pull request by mwest1066**
To-do items:
- [x] Decide whether to keep the `authn_providers` table or just use strings [decision: keep it].
- [x] Decide whether to remove `versionString` from tokens [decision: removed it].
- [x] Decide how to handle LTI [decision: current code in this PR maintains functionality; defer improvements to later].
- [x] Write tests for default user-in-course-institution allowAccess rules.
- [x] Should we change the Default institution to Public? [decision: not now; might reconsider later]

This PR does several things:

1. Creates the `authn_providers` table, which currently includes "Shibboleth", "Google", "Azure", and "LTI". Having a table for this might be overkill, and maybe we could just use plain strings for each provider, rather than making a table.

2. Creates the `institution_authn_providers` table, which lists the allowed authentication providers for each institution. This handles the case of UBC where they only want to permit Shib authn.

3. Changes institution uid patterns to use regexps instead of SQL LIKE. This is just future-proofing so we can do things like `@uiuc.edu$|@illinois.edu$` (matching two different UID patterns).

4. Forces each course to have a not-NULL `institution_id`.

5. Adds `users.institution_id` (not NULL), so each user is assigned to an institution. This happens at login. This will make it easier in the future to show admins all users in each institution, and things like that.

6. Add a `versionString` to our client-cookie-token code (starts at `1`). Incrementing this version causes all users to be logged out (their `pl_authn` cookies will no longer match), forcing them to re-authenticate. This was done to force everyone to re-auth to get the `institution_id` set correctly, but now I'm thinking this wasn't necessarily a good idea, because this `versionString` will make blue/green deployments harder. Maybe we should just add a "force-institution-id-update" button for admins.

7. To handle LTI, I made both an "LTI" institution and an "LTI" provider. @trombonekenney: I'd appreciate feedback on whether this is the right way to do things. When we authn an LTI user, we assign their `institution_id` to the LTI institution. We also set `locals.authn_provider_name = "LTI"`, which is the variable we use in EJS to do things like block course adds. I'd be very happy to change this to some other system if you'd prefer.

8. I made it so that courseInstance allowAccess rules default to restricting access to the users in the same institution as the course. This makes all courses default to working like our UIUC courses do now.

After review and possible updates to the above points (especially LTI), I plan to merge this and deploy the new master. This will automatically create the default institution (id=1) and assign all courses and users to it. Then I will run SQL like the following to update to our current state:

```
INSERT INTO institutions (short_name, long_name, uid_regexp)
VALUES
('UIUC', 'University of Illinois at Urbana-Champaign', '@illinois\.edu$'),
('ZJUI', 'Zhejiang University/UIUC Institute', '@intl\.zju\.edu\.cn$'),
('gvsu.edu', 'Grand Valley State University', '@(mail\.|)gvsu\.edu$')
;

UPDATE pl_courses
SET institution_id = i.id
FROM institutions AS i
WHERE i.short_name = 'UIUC';

UPDATE users AS u
SET institution_id = i.id
FROM institutions AS i
WHERE u.uid ~ i.uid_regexp;
```

### CODE REVIEW

**Comment by tbretl** [comment]
This all seems ok at a quick glance.

I am a little worried about the choice (perhaps made in the PR that introduced the `institution`) to grant access to institutions by including a magic string - a name, abbreviation, internet domain, etc., that represents the institution - in an `accessRule` in `infoCourseInstance.json`. It seems like this could get annoying as the number of institutions grows, or as preferred names of institutions change (e.g., UIUC vs Illinois). It also seems easy to accidentally grant access to the wrong institution. Institutions are different from roles (for example) in this respect - roles are not likely to change much, and the number of them is likely to remain small.

One alternative would be to do institutional access control in the browser, with a UI that interacts directly with the DB instead of with JSON that gets synced to the DB.

**Comment by coatless** [comment]
Future-wise, it might be more convenient to adopt the [Seamless Access](https://seamlessaccess.org/) standard instead of e-mail based regex.

**Comment by trombonekenny** [comment]
> This all seems ok at a quick glance.
> 
> I am a little worried about the choice (perhaps made in the PR that introduced the `institution`) to grant access to institutions by including a magic string - a name, abbreviation, internet domain, etc., that represents the institution - in an `accessRule` in `infoCourseInstance.json`. It seems like this could get annoying as the number of institutions grows, or as preferred names of institutions change (e.g., UIUC vs Illinois). It also seems easy to accidentally grant access to the wrong institution. Institutions are different from roles (for example) in this respect - roles are not likely to change much, and the number of them is likely to remain small.
> 
> One alternative would be to do institutional access control in the browser, with a UI that interacts directly with the DB instead of with JSON that gets synced to the DB.

We updated the adminOverview page to include a list of institutions to help with this. I presume we could make a similar page ServerProperties (?) that gives a read-only view of different PL site-specific configurations from the DB. Then we'd at least have an user-viewing authoritative source to check against?

File: migrations/156_users__institution_id__add.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE users ADD COLUMN institution_id BIGINT NOT NULL DEFAULT 1 REFERENCES institutions(id) ON UPDATE CASCADE ON DELETE SET NULL;
+ALTER TABLE users DROP COLUMN provider;
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply migrations/156_users__institution_id__add.sql]: Good question! The only thing we used this data for was controlling the logout button display and action. We now handle that with `locals.authn_provider_name`, which is stored in the `pl_authn` cookie.

However, there is a slight issue during upgrade where this variable won't be set in the cookie. I've just added some code to force a re-login in this case.

**Comment by mwest1066** [comment]
> > This all seems ok at a quick glance.
> > I am a little worried about the choice (perhaps made in the PR that introduced the `institution`) to grant access to institutions by including a magic string - a name, abbreviation, internet domain, etc., that represents the institution - in an `accessRule` in `infoCourseInstance.json`. It seems like this could get annoying as the number of institutions grows, or as preferred names of institutions change (e.g., UIUC vs Illinois). It also seems easy to accidentally grant access to the wrong institution. Institutions are different from roles (for example) in this respect - roles are not likely to change much, and the number of them is likely to remain small.
> > One alternative would be to do institutional access control in the browser, with a UI that interacts directly with the DB instead of with JSON that gets synced to the DB.
> 
> We updated the adminOverview page to include a list of institutions to help with this. I presume we could make a similar page ServerProperties (?) that gives a read-only view of different PL site-specific configurations from the DB. Then we'd at least have an user-viewing authoritative source to check against?

I'm certainly happy to revisit the method by which we control institution access. For now we are just keeping it unchanged.

I agree that it would be very helpful to soon add a list of valid institutions somewhere that instructors can see it. I'll add an issue for this.

**Pull request merged**